components/ContainedInputViews/src/MDCContainedInputView.h

```diff
@@ -0,0 +1,319 @@
+// limitations under the License.
+
+#import <Foundation/Foundation.h>
+#import <UIKit/UIKit.h>
+
+/**
+ A set of Contained Input View states outlined in the Material guidelines. These states overlap with
+ and extend UIControlState.
+ */
+typedef NS_OPTIONS(NSInteger, MDCContainedInputViewState) {
+  /**
+   The default state of the contained input view.
+   */
+  MDCContainedInputViewStateNormal = 1 << 0,
+  /**
+   The state the view is in during normal editing.
+   */
+  MDCContainedInputViewStateFocused = 1 << 1,
+  /**
+   This state most closely resembles the @c selected UIControlState.
+   */
+  MDCContainedInputViewStateActivated = 1 << 2,
+  /**
+   The error state.
+   */
+  MDCContainedInputViewStateErrored = 1 << 3,
+  /**
+   The disabled state.
+   */
+  MDCContainedInputViewStateDisabled = 1 << 4,
+};
+
+/**
+ Dictates the relative importance of the underline labels, and the order in which they are laid out.
+ */
+typedef NS_ENUM(NSUInteger, MDCContainedInputViewUnderlineLabelDrawPriority) {
+  /**
+   When the priority is @c .leading, the @c leadingUnderlineLabel will be laid out first within the
+   horizontal space available for @b both underline labels. Any remaining space will then be given
+   for the @c trailingUnderlineLabel.
+   */
+  MDCContainedInputViewUnderlineLabelDrawPriorityLeading,
+  /**
+   When the priority is @c .trailing, the @c trailingUnderlineLabel will be laid out first within
+   the horizontal space available for @b both underline labels. Any remaining space will then be
+   given for the @c leadingUnderlineLabel.
+   */
+  MDCContainedInputViewUnderlineLabelDrawPriorityTrailing,
+  /**
+   When the priority is @c .custom, the @c customUnderlineLabelDrawPriority property will be used to
+   divide the space available for the two underline labels.
+   */
+  MDCContainedInputViewUnderlineLabelDrawPriorityCustom,
+};
+
+/**
+ This enum represents different states the floating label can be in.
+ */
+typedef NS_ENUM(NSUInteger, MDCContainedInputViewFloatingLabelState) {
+  /**
+   The state where the floating label is not visible.
+   */
+  MDCContainedInputViewFloatingLabelStateNone,
+  /**
+   The state where the floating label is floating.
+   */
+  MDCContainedInputViewFloatingLabelStateFloating,
+  /**
+   The state where the floating label is occupying the normal text area.
+   */
+  MDCContainedInputViewFloatingLabelStateNormal,
+};
+
+@protocol MDCContainedInputViewStyle;
+@protocol MDCContainedInputViewColorScheming;
+
+@protocol MDCContainedInputView <NSObject>
+/**
+ Dictates the @c MDCContainerStyle of the text field. Defaults to an instance of
+ MDCContainerStyleBase.
+ */
+@property(nonatomic, strong, nonnull) id<MDCContainedInputViewStyle> containerStyle;
+
+/**
+ Describes the current @c MDCContainedInputViewState of the view.
+ */
+@property(nonatomic, assign, readonly) MDCContainedInputViewState containedInputViewState;
```

Comment:
Clarification: 
How does containedInputViewState correlates to UIControl's "state"? 
How do you expect classes that conform to this protocol to make sure the two align?

Reply:
It coexists with UIControlState. UIControlState affects the value returned by the computed MDCContainedInputViewState property, but not the other way around. MDCContainedInputViewState has states not accounted for by UIControlState, like error.

The approach I took with the textfield, for example, ensured that UIControlState behavior was unchanged.

If we had an enum or option set that represented these same Material states it might be reasonable to use that instead of introduce a new type. We don't have something like that though.